### DISCUSSION TITLE
How to Suppress PMD Warnings for Code Duplicacy in C++ Lookup Tables?

### DISCUSSION THREAD

**Discussion by PragatiGarg-eaton** [pmd/pmd#5294] [Q&A] [answered]
Hi everyone,

I'm currently working on a C++ project that involves extensive use of lookup tables. While these tables are essential for the functionality of my application, I'm encountering PMD warnings about code duplicacy. I understand that these warnings are meant to help identify redundant code, but in this case, the duplicacy is intentional and necessary for the lookup tables to function correctly.

I would appreciate any advice or best practices on how to manage and suppress these PMD warnings effectively. Here are a few specific questions I have:

**Suppressing Warnings**: What are the best methods to suppress PMD warnings in C++? I've read about using `// NOPMD` comment. But it is not working effectively.
I have gone through https://docs.pmd-code.org/latest/pmd_userdocs_suppressing_warnings.html documentation also, but it is not helping.

**Sample of look up table**: 
`// Suppress PMD warning for code duplicacy
// NOPMD
const BIT_LIST_LOOKUP[x] = // NOPMD
{
	BIT_LIST_LOOKUP_INVALID_BIT,
	0U, 1U, 0U, 1U, 0U, 1U, 0U,
	1U, 0U, 1U, 0U, 1U, 0U, 1U, 0U,
	1U, 0U, 1U, 0U, 1U, 0U, 1U, 0U,
	1U, 0U, 1U, 0U, 1U, 0U, 1U, 0U
};// TURN_OFF_WARNINGS`

Any insights, examples, or resources you could share would be greatly appreciated. Thank you in advance for your help!

**Comment by jsotuyod** [comment, marked as answer]
Hi. You are looking at the PMD suppression doc, but should be looking at the CPD specific reference at https://docs.pmd-code.org/latest/pmd_userdocs_cpd.html#suppression

CPD uses a specific `CPD-OFF` / `CPD-ON` comment pair that easily allows to ignore arbitrary blocks of code.

**Reply by PragatiGarg-eaton** [reply]
Thanks for the quick help! It's working perfectly.
